bpo-28002: Roundtrip f-strings with ast.unparse better

hauntsaninja opened the pull request:
By attempting to avoid backslashes in f-string expressions.
We also now proactively raise errors for some backslashes we can't
avoid while unparsing FormattedValues

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-28002](https://bugs.python.org/issue28002) -->
https://bugs.python.org/issue28002
<!-- /issue-number -->

Automerge-Triggered-By: GH:isidentical

Commit:
bpo-28002: Roundtrip f-strings with ast.unparse better

By attempting to avoid backslashes in f-string expressions.
We also now proactively raise errors for some backslashes we can't
avoid while unparsing FormattedValues

isidentical commented on Lib/ast.py:
```diff
@@ -1018,15 +1020,39 @@ def visit_AsyncWith(self, node):
+                if quote_type not in value:
+                    self.write(f"{quote_type}{value}{quote_type}")
+                    return
+        self.write(repr(value))
```
What do you think about simplifying this a bit with something like this;
```suggestion
        use_repr = any(map(should_use_repr, value))

        if not use_repr:
            quote_types = ["'", '"', '"""', "'''"]

            if "\n" in value:
                quote_types = quote_types[2:]

            for quote_type in quote_types:
                if quote_type not in value:
                    self.write(f"{quote_type}{value}{quote_type}")
                    break
            else:
                use_repr = True

        if use_repr:
            self.write(repr(value))
```

hauntsaninja replied:
Sure! That's more explicit about the control flow.

hauntsaninja committed:
Update Lib/ast.py

Co-authored-by: Batuhan Taskaya <isidentical@gmail.com>

isidentical commented on Lib/test/test_unparse.py:
```diff
@@ -475,7 +490,6 @@ class DirectoryTestCase(ASTTestCase):
     lib_dir = pathlib.Path(__file__).parent / ".."
     test_directories = (lib_dir, lib_dir / "test")
-    skip_files = {"test_fstring.py"}
     run_always_files = {"test_grammar.py", "test_syntax.py", "test_compile.py",
```
Can you add `test_fstring.py` to the  `run_always_files`, so we can ensure it is always runned.

isidentical commented on Lib/ast.py:
```diff
@@ -646,7 +647,8 @@ class _Unparser(NodeVisitor):
     is disregarded."""
 
-    def __init__(self):
+    def __init__(self, avoid_backslashes=False):
```
Would be nice to also have this keyword-only

Commit:
_Unparser: make avoid_backslashes keyword only

Commit:
test_unparse: add test_fstring to run_always_files

Commit:
unparse fstrings: improve cosmetics by escaping whitespace

Commit:
unparse: add another test case

Commit:
unparse: fix final quote escaping issues

The previous cosmetic fix means we use triple quotes less often, so this
is less likely to occur in practice, but should be fixed anyway.

Commit:
unparse: simplify, share code with docstring writing

Commit:
Merge remote-tracking branch 'upstream/master' into astunparse

Commit:
remove comment

isidentical commented on Lib/ast.py:
```diff
@@ -1097,33 +1166,19 @@ def visit_Name(self, node):
-        self.write(f'"""{value}"""')
+        value, quote_types = self._str_literal_helper(node.value, ('"""', "'''"))
+        quote_type = quote_types[0]
+        self.write(f"{quote_type}{value}{quote_type}")
```
I feel like this is extremely similar to `_write_str_avoiding_backslashes`. Would you mind to use it here (and add a `**kwargs` parameter to pass there and pass it to helper)

hauntsaninja replied:
Thanks, changed!

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1055,80 @@ def visit_AsyncWith(self, node):
+            c for c in value
+            if c not in ' \n\t' and unicodedata.category(c)[0] in ("C", "Z")
+        }
+        qts = quote_types
```
I don't think this is necessary, in both cases you are defining `qts = [<comprehension>]` below.

hauntsaninja replied:
Changed, but note we now might do the (relatively) more expensive `in` filtering on quote types we wouldn't have to.

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1055,80 @@ def visit_AsyncWith(self, node):
+        qts = quote_types
+        val = value.replace("\\", "\\\\")
+        for c in escape:
+            val = val.replace(c, c.encode('unicode_escape').decode('ascii'))
```
I don't think replacing is the suitable solution for this, you should use `var[x] = blabla` or simply `val = "".join(char.encode(x).decode(y) for char in val).replace("\\", "\\\\")`...

hauntsaninja replied:
Changed to use join! ~We need to replace backslashes before escaping, though.~ (since we're now going character by character as opposed to escape by escape we can just add backslash to the characters that need escaping)

isidentical reviewed (commented):
The general approach seems reasonable but I believe there are still parts that can be improvable in code quality wise. Thank you so much on your work on this

Commit:
reduce assignments to qts

Commit:
use join instead of multiple replace

Commit:
use write_str_avoiding_backslashes for docstrings

Commit:
don't use replace for backslashes

Commit:
use isprintable

I didn't know this existed, since I was just tracing the C code for
repr. Also must have missed it on google

Commit:
fix comment

isidentical commented on Lib/ast.py:
```diff
@@ -653,7 +653,8 @@ class _Unparser(NodeVisitor):
 
-    def __init__(self):
+    def __init__(self, *, avoid_backslashes=False):
+        self.avoid_backslashes = avoid_backslashes
```
```suggestion
    def __init__(self, *, _avoid_backslashes=False):
        self._avoid_backslashes = _avoid_backslashes
```

ericvsmith replied:
What's the logic behind that change? I'm not sure that's an improvement.

isidentical replied:
> What's the logic behind that change? I'm not sure that's an improvement.

In the future, this interface might be opened up to public (`_Unparser`) and I don't think this should be any different than `_indent`, they both represent a state for the unparsing operation. And this one especially can be removed if we find a nicer way.

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,77 @@ def visit_AsyncWith(self, node):
 
+    def _str_literal_helper(
+        self, value, quote_types=("'", '"', '"""', "'''"), escape=""
+    ):
```
(just a preference, feel free to ignore it)
```suggestion
    def _str_literal_helper(
        self, string, quote_types=("'", '"', '"""', "'''"), _custom_escapes=()
    ):
```

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,77 @@ def visit_AsyncWith(self, node):
+    ):
+        """Helper for writing string literals, minimising escapes.
+        Returns (possible quote types, string literal to write).
+        """
```
I don't get the reason for returning a list of quotes instead of the most suitable one. As far as I can see, both usages just accesses the first item so we may just return the most suiting quote
```suggestion
        """Helper for writing string literals, minimising escapes. Returns the
        possible quote type, and the escaped valued."""
```

ericvsmith replied:
I think somewhere we've agreed on American English spellings, so I'd make that "minimizing".

hauntsaninja replied:
In `visit_JoinedStr` we actually repeatedly filter the list of possibly quote types within the loop, so we can't just return a single quote.

I'll make the American English change.

hauntsaninja replied:
Added a comment to `visit_JoinedStr` to help make that stand out.

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,77 @@ def visit_AsyncWith(self, node):
+            c.encode('unicode_escape').decode('ascii')
+            if c in escape or (not c.isprintable() and c not in '\n\t') else c
+            for c in value
+        )
```
This looks like a very long single expression, what about turning this into a function just like this;
```suggestion
        def escape_char(char):
            if char in (*_custom_escapes, "\\") or (
                not char.isprintable() and char not in "\n\t"
            ):
                return char.encode("unicode_escape").decode("ascii")
            else:
                return char
        escaped_string = "".join(map(escape_char, string))
```

ericvsmith replied:
I think the original is clearer, myself, mostly because it avoids `map()`.

isidentical replied:
Initially, I failed to understand the whole thing since the designated result from the if-expr was in the first line (since it doesn't fit to 79 chars). Just mapping to a single function and giving the returned iterator to something that consumes iterables is something I found reasonable, and IMHO the seperation of the if-exp to its own function is just makes things self-explaning.

hauntsaninja replied:
I parenthesised and indented the ternary expression. I think this strikes a good balance, but happy to change it if you still feel a function is better.

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,77 @@ def visit_AsyncWith(self, node):
+        )
+        qts = [quote for quote in quote_types if quote not in val]
+        if "\n" in val:
+            qts = [quote for quote in qts if quote in ('"""', "'''")]
```
Please use descriptive names, such as `possible_quotes`
```suggestion
        possible_quotes = [quote for quote in quote_types if quote not in escaped_string]
        if "\n" in escaped_string:
            # For multi-line strings, use triple quotes
            possible_quotes = [
                quote for quote in possible_quotes if len(quote) == 3
            ]
```

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,77 @@ def visit_AsyncWith(self, node):
+            # on the original value. Try to use a quote_type from quote_types.
+            value = repr(value)
+            quote_type = next((q for q in quote_types if value[0] in q), value[0])
+            return value[1:-1], [quote_type]
```
Would be nice to move this function's body to an inner function called `fallback` and return it there, so we can replace the `assert len(qts[0]) == 3` check with this
```
        def fallback(original):
            quote, *escaped_string_parts, quote = repr(original)
            for possible_quote in quote_types:
                if quote in possible_quote:
                    break
            else:
                possible_quote = quote
            return "".join(escaped_string_parts), possible_quote
```

hauntsaninja replied:
(Resolving this, since I replied on the other related comment https://github.com/python/cpython/pull/19612#discussion_r436394009)

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,77 @@ def visit_AsyncWith(self, node):
+            qts.sort(key=lambda q: q[0] == val[-1])
+            # If we're using triple quotes and we'd need to escape a final quote, escape
+            if qts[0][0] == val[-1]:
+                assert len(qts[0]) == 3
```
Replace this assert with an if, and if the result is false call fallback. We dont want assertion errors to pop out if we can traverse that string

hauntsaninja replied:
The assert will always pass (as long as people pass valid quotes to quote_types, which is okay to assume since this is an internal helper, if not, we should just enforce that they do). This is because we'd have filtered out the single quote that we need to escape already, since it's clearly in the string.

If this is confusing, I can remove the assert (although I like it, since it helps clarify the comment), or raise another kind of exception — but we don't need to fallback for a condition that will never trigger.

isidentical reviewed (commented):
(do not solely apply suggestions, please fix other references to changed variable / attribute names etc.) Thanks for all the effort, it looks good overall but these suggestions can increase the quality and help us for future maintenance. I'll do a scan of 4000~ top PyPI packages when this PR is final, as far as I can tell this might be the last PR that will touch the `ast.unparse` for 3.9 so it is better to be prepared :)

Commit:
Use American English

Commit:
s/qts/possible_quotes/g

Commit:
Prefix undesrscore to avoid_backslashes

Commit:
Rename value and val

Commit:
Reformat string escaping

Commit:
Add a comment to visit_JoinedStr

Commit:
Fix line lengths

Commit:
Update comment for variable name change

isidentical commented on Lib/ast.py:
```diff
@@ -653,7 +653,8 @@ class _Unparser(NodeVisitor):
 
-    def __init__(self):
+    def __init__(self, *, _avoid_backslashes=False):
+        self._avoid_backslashes = _avoid_backslashes
```
Please move this to the end of the `__init__` (right after `_precedences`)

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,82 @@ def visit_AsyncWith(self, node):
 
+    def _str_literal_helper(
+        self, string, quote_types=("'", '"', '"""', "'''"), escape=""
+    ):
```
There are too many different constants hanging around (all quote types, only multiple quotes etc.), let's make them private globals;
```diff
+_SINGLE_QUOTES = ("'", '"')
+_MULTI_QUOTES = ('"""', "'''")
+_ALL_QUOTES = (*_SINGLE_QUOTES, *_MULTI_QUOTES)
+
 class _Unparser(NodeVisitor):
```

isidentical replied:
I spent some time on it about how we could possibly simplify this interface. From what I see `escape` can be either `"\n\t"` or an empty string, which allows us to reshape the signature like the following;
```diff
     def _str_literal_helper(
-        self, string, quote_types=("'", '"', '"""', "'''"), escape=""
+        self,
+        string,
+        *,
+        escape_special=False,
+        quote_types=_ALL_QUOTES,
     ):
```
The `escape_special` should control the behavior of escape of `\n\t`.

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,82 @@ def visit_AsyncWith(self, node):
+        self, string, quote_types=("'", '"', '"""', "'''"), escape=""
+    ):
+        """Helper for writing string literals, minimizing escapes.
+        Returns (possible quote types, string literal to write).
```
Even though it is implicitly said with parens, it would be nice to clarify the return value is a 'tuple' of those.

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,82 @@ def visit_AsyncWith(self, node):
+        """
+        # Escape characters we've been told to escape, backslashes, and
+        # non-printable characters other than \n and \t.
+        escape = {*escape, '\\'}
```
Since the signature might get simplified, no longer need this.

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,82 @@ def visit_AsyncWith(self, node):
+                else c
+            )
+            for c in string
+        )
```
As I said in the previous review, I still find this part very hard to read. Like `esc_char` you removed below, I think it would be best if we can make it a function with 'pure' conditions.
```diff
+        def escape_char(char):
+            if not escape_special and char in "\n\t":
+                return char
+            elif char == "\\" or not char.isprintable():
+                return char.encode("unicode_escape").decode("ascii")
+            else:
+                return char
+        escaped_string = "".join(map(escape_char, string))
```

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,82 @@ def visit_AsyncWith(self, node):
+        )
+        possible_quotes = [q for q in quote_types if q not in escaped_string]
+        if "\n" in escaped_string:
+            possible_quotes = [q for q in possible_quotes if q in ('"""', "'''")]
```
Is it possible that we could somehow merge these 2 different iterations? (something like this?)
```diff
+        is_multi_line = "\n" in escaped_string
+
+        possible_quote_types = quote_types
+        if is_multi_line:
+            possible_quote_types = _MULTI_QUOTES
+        possible_quotes = [quote for quote in possible_quote_types if quote not in escaped_string]
```

hauntsaninja replied:
The suggestion isn't quite correct, since we'd need to make sure `possible_quote_types` is a subset of `quote_types`. We could do something like:
```
is_multi_line = "\n" in escaped_string
possible_quotes = [q for q in quote_types if (q not in escaped_string) and (not is_multi_line or q in _MULTI_QUOTES)]
```
But I think that's less readable. `len(quote_types) <= 4`, so removing an iteration isn't important.

isidentical replied:
> we'd need to make sure possible_quote_types is a subset of quote_types

Makes sense, what about?
```diff
+        is_multi_line = "\n" in escaped_string
+
+        possible_quote_types = quote_types
+        if is_multi_line:
+            possible_quote_types = [quote for quote in possible_quote_types if quote in _MULTI_QUOTES]
+        possible_quotes = [quote for quote in possible_quote_types if quote not in escaped_string]
```

hauntsaninja replied:
Sure, I'll swap the order of the iterations. That could save us two iterations over the escaped string.

isidentical commented on Lib/ast.py:
```diff
@@ -1053,15 +1054,82 @@ def visit_AsyncWith(self, node):
+            # on the original string. Try to use a quote from quote_types.
+            string = repr(string)
+            quote = next((q for q in quote_types if string[0] in q), string[0])
+            return string[1:-1], [quote]
```
This `next()` is a bit confusing. Is it for using the triple-quote for of whether double or single quote used? If so, I guess we can refactor it to something a bit more simpler?
```diff
+            original = repr(string)
+            quote_ty = original[0]
+            return original[1:-1], [quote_ty * 3 if is_multi_line else quote_ty]
```

hauntsaninja replied:
This doesn't quite do what we want. E.g., we use this code for docstrings, which we want to be triple quotes even if there aren't newlines. I added that to the comment though :-)

isidentical reviewed (commented):
A final review, more about some simplifications and little refactors!

Commit:
move variable from start of __init__ to end

Commit:
use global constants for quote types

Commit:
make docstring explicit about returning a tuple

Commit:
use kw only args

Commit:
add escape_special_whitespace boolean argument

Commit:
use an inner function instead of a comprehension

Commit:
add to comment when falling back to repr

isidentical reviewed (approved):
@pablogsal would you mind checking out this?

Commit:
swap the order of the iterations to save a little effort

ericvsmith reviewed (commented):
Sorry for the delay. This looks reasonable enough to me, but someone who's more familiar with the AST and/or unparsing should take a look. Maybe @ambv?
